File: sth/mill/report/laporan_penerimaan_tbs_mill/laporan_penerimaan_tbs_mill.js
// Copyright (c) 2024, Your Company and contributors
// For license information, please see license.txt

frappe.query_reports["Laporan Penerimaan TBS Mill"] = {
	"filters": [
		// {
		// 	"fieldname": "laporan",
		// 	"label": __("Laporan"),
		// 	"fieldtype": "Select",
		// 	"options": ["Laporan Penerimaan TBS", "Rekap Penerimaan TBS"],
		// 	"default": "Laporan Penerimaan TBS",
		// 	"reqd": 1,
		// 	"on_change": function () {
		// 		let report = frappe.query_report;
		// 		let laporan_type = report.get_filter_value('laporan');

		// 		// Show/hide filters based on report type
		// 		if (laporan_type === "Rekap Penerimaan TBS") {
		// 			report.get_filter('tbs').toggle(false);
		// 			report.get_filter('supplier').toggle(false);
		// 			report.get_filter('tanggal_dari').toggle(false);
		// 			report.get_filter('tanggal_sampai').toggle(false);
		// 			report.get_filter('company').toggle(true);
		// 			report.get_filter('tanggal').toggle(true);
		// 		} else {
		// 			report.get_filter('tbs').toggle(true);
		// 			report.get_filter('supplier').toggle(true);
		// 			report.get_filter('tanggal_dari').toggle(true);
		// 			report.get_filter('tanggal_sampai').toggle(true);
		// 			report.get_filter('company').toggle(false);
		// 			report.get_filter('tanggal').toggle(false);
		// 		}

		// 		report.refresh();
		// 	}
		// },
		{
			"fieldname": "tbs",
			"label": __("TBS"),
			"fieldtype": "Select",
			"options": ["External", "Internal"],
			"default": "",
			"reqd": 0
		},
		{
			"fieldname": "supplier",
			"label": __("Supplier"),
			"fieldtype": "Link",
			"options": "Supplier",
			"reqd": 0,
			"get_query": function () {
				return {
					"filters": {
						"disabled": 0
					}
				}
			}
		},
		{
			"fieldname": "divisi",
			"label": __("Divisi"),
			"fieldtype": "Link",
			"options": "Divisi",
		},
		{
			"fieldname": "tanggal_dari",
			"label": __("Tanggal Dari"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_start(),
			"reqd": 0
		},
		{
			"fieldname": "tanggal_sampai",
			"label": __("Tanggal Sampai"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_end(),
			"reqd": 0
		},
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 0,
		},
		{
			"fieldname": "unit",
			"label": __("Unit"),
			"fieldtype": "Link",
			"options": "Unit",  // Replace with your actual doctype name
			"reqd": 0,
			"get_query": function () {
				let company = frappe.query_report.get_filter_value('company');
				let filters = {
					"mill": 1
				};

				if (company) {
					filters["company"] = company;
				}

				return {
					"filters": filters
				}
			}

		},

		{
			"fieldname": "tanggal",
			"label": __("Tanggal"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 0,
			"hidden": 1
		}
	],

	formatter: function (value, row, column, data, default_formatter) {
		const number_fields = [
			"bruto",
			"tara",
			"netto_1",
			"sort",
			"netto_2",
			"bjr",
			"grader"
		];

		if (number_fields.includes(column.fieldname) && value != null) {
			value = frappe.format(value, {
				fieldtype: "Float",
				precision: 0
			});
		}

		value = default_formatter(value, row, column, data);

		if (data && data.is_group == 1) {
			value = `<b>${value}</b>`;
		}

		return value;
	}

	// "formatter": function (value, row, column, data, default_formatter) {
	// 	value = default_formatter(value, row, column, data);

	// 	let laporan_type = frappe.query_report.get_filter_value('laporan');

	// 	if (laporan_type === "Rekap Penerimaan TBS") {
	// 		// Format for Rekap report
	// 		if (column.fieldname !== "driver_name") {
	// 			if (value && !isNaN(parseFloat(value))) {
	// 				value = `<div style="text-align: right">${parseFloat(value).toLocaleString('id-ID', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}</div>`;
	// 			}
	// 		}

	// 		// Bold formatting for total rows
	// 		if (data && (data.driver_name === "Total Internal" || data.driver_name === "Total External")) {
	// 			value = `<div style="font-weight: bold">${value}</div>`;
	// 		}
	// 	} else {
	// 		// Format for Laporan Penerimaan TBS
	// 		if (column.fieldname == "bruto" ||
	// 			column.fieldname == "tara" ||
	// 			column.fieldname == "netto" ||
	// 			column.fieldname == "potongan" ||
	// 			column.fieldname == "berat_normal") {
	// 			if (value && !isNaN(parseFloat(value))) {
	// 				value = `<div style="text-align: right">${parseFloat(value).toFixed(2)}</div>`;
	// 			}
	// 		}
	// 	}

	// 	return value;
	// },

	// "onload": function (report) {
	// 	// Initialize filter visibility
	// 	let laporan_type = report.get_filter_value('laporan');

	// 	if (laporan_type === "Rekap Penerimaan TBS") {
	// 		report.get_filter('tbs').toggle(false);
	// 		report.get_filter('supplier').toggle(false);
	// 		report.get_filter('tanggal_dari').toggle(false);
	// 		report.get_filter('tanggal_sampai').toggle(false);
	// 		report.get_filter('company').toggle(true);
	// 		report.get_filter('tanggal').toggle(true);
	// 	} else {
	// 		// report.get_filter('company').toggle(false);
	// 		report.get_filter('tanggal').toggle(false);
	// 	}
	// }
};

File: sth/mill/report/laporan_penerimaan_tbs_mill/laporan_penerimaan_tbs_mill.py
# Copyright (c) 2026, DAS and contributors
# For license information, please see license.txt

from frappe import _
import frappe
from collections import defaultdict
from frappe.utils import flt

def execute(filters=None):
	laporan_type = filters.get("tbs", "All")
	columns = get_columns()
	data = []

	if laporan_type == "External":
		report_rows = get_data_external(filters)
	elif laporan_type == "Internal":
		report_rows = get_data_internal(filters)
	else:  # "All"
		report_rows = get_data_all(filters)

	for row in report_rows:
		data.append({
			"is_group": 1,
			"no_polisi": row["grouping"]
		})

		for child in row["data"]:
			child["is_group"] = 0
			data.append(child)

		data.append({
			"is_group": 1,
			"jam_keluar": "Total",
			"bruto": round(sum(flt(d["bruto"] or 0) for d in row["data"]), 0),
			"tara": round(sum(flt(d["tara"] or 0) for d in row["data"]), 0),
			"netto_1": round(sum(flt(d["netto_1"] or 0) for d in row["data"]), 0),
			"sort": round(sum(flt(d["sort"] or 0) for d in row["data"]), 0),
			"netto_2": round(sum(flt(d["netto_2"] or 0) for d in row["data"]), 0),
			"bjr": round(sum(flt(d["bjr"] or 0) for d in row["data"]), 0),
			"grader": round(sum(flt(d["grader"] or 0) for d in row["data"]), 0),
		})

	return columns, data

def get_columns():
	"""Define kolom-kolom untuk laporan penerimaan TBS External"""
	return [
		{
			"fieldname": "no_polisi",
			"label": _("No. Polisi"),
			"fieldtype": "Data",
			"width": 100
		},
		{
			"fieldname": "no_referensi",
			"label": _("No. Referensi"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "no_slip_timbang",
			"label": _("No. Slip Timbang"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "nama_supir",
			"label": _("Nama Supir"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "tgl_masuk",
			"label": _("Tanggal Masuk"),
			"fieldtype": "Date",
			"width": 150
		},
		{
			"fieldname": "jam_masuk",
			"label": _("Jam Masuk"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "tgl_keluar",
			"label": _("Tanggal Keluar"),
			"fieldtype": "Date",
			"width": 150
		},
		{
			"fieldname": "jam_keluar",
			"label": _("Jam Keluar"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "bruto",
			"label": _("Bruto (Kg)"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "tara",
			"label": _("Tara (Kg)"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "netto_1",
			"label": _("Netto I (Kg)"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "sort",
			"label": _("Sort (Kg)"),
			"fieldtype": "Data",
			"width": 150
		},
  	{
			"fieldname": "netto_2",
			"label": _("Netto II (Kg)"),
			"fieldtype": "Data",
			"width": 150
		},
  	{
			"fieldname": "bjr",
			"label": _("BJR"),
			"fieldtype": "Data",
			"width": 150
		},
  	{
			"fieldname": "grader",
			"label": _("Grader"),
			"fieldtype": "Data",
			"width": 150
		},
	]

def get_condition(filters, report_type):
	"""report_type dikirim eksplisit oleh caller: "External" / "Internal" / "All" —
	supaya filter supplier/divisi diterapkan dengan benar sesuai konteks,
	tidak lagi bergantung pada filters.get("tbs")."""
	conditions = ""

	if filters.get("supplier") and report_type in ("External", "All"):
		conditions += " AND t.supplier = %(supplier)s"

	if filters.get("divisi") and report_type in ("Internal", "All"):
		conditions += " AND spb.divisi = %(divisi)s"

	if filters.get("tanggal_dari") and filters.get("tanggal_sampai"):
		conditions += " AND t.posting_date BETWEEN %(tanggal_dari)s AND %(tanggal_sampai)s"

	if filters.get("unit"):
		conditions += " AND t.unit = %(unit)s"

	if filters.get("company"):
		conditions += " AND t.company = %(company)s"

	return conditions

def get_data_external(filters):
	conditions = get_condition(filters, "External")
	sql = frappe.db.sql("""
		SELECT
			t.name,
			s.supplier_name as grouping,
			t.no_polisi,
			t.ticket_number as no_referensi,
			t.name as no_slip_timbang,
			t.driver_name as nama_supir,
			t.posting_date as tgl_masuk,
			t.weight_in_time as jam_masuk,
			t.posting_date as tgl_keluar,
			t.weight_out_time as jam_keluar,
			FORMAT(t.bruto, 2) as bruto,
			FORMAT(t.tara, 2) as tara,
			FORMAT(t.netto, 2) as netto_1,
			FORMAT(t.netto * (t.potongan_sortasi / 100), 2) as sort,
			FORMAT(t.netto_2, 2) as netto_2,
			FORMAT(t.isi_komidel, 2) as bjr,
			FORMAT(0, 2) as grader
		FROM `tabTimbangan` t
		LEFT JOIN `tabSupplier` s ON s.name = t.supplier
		WHERE t.receive_type = "TBS Eksternal" AND t.supplier IS NOT NULL AND t.docstatus = 1 {};
	""".format(conditions), filters, as_dict=True)

	return _group_rows(sql)

def get_data_internal(filters):
	conditions = get_condition(filters, "Internal")
	sql = frappe.db.sql("""
		SELECT
			t.name,
			spb.divisi as grouping,
			t.no_polisi,
			t.ticket_number as no_referensi,
			t.name as no_slip_timbang,
			t.driver_name as nama_supir,
			t.posting_date as tgl_masuk,
			t.weight_in_time as jam_masuk,
			t.posting_date as tgl_keluar,
			t.weight_out_time as jam_keluar,
			FORMAT(t.bruto, 2) as bruto,
			FORMAT(t.tara, 2) as tara,
			FORMAT(t.netto, 2) as netto_1,
			FORMAT(t.netto * (t.potongan_sortasi / 100), 2) as sort,
			FORMAT(t.netto_2, 2) as netto_2,
			FORMAT(t.isi_komidel, 2) as bjr,
			FORMAT(0, 2) as grader
		FROM `tabTimbangan` t
		LEFT JOIN `tabSurat Pengantar Buah` as spb ON spb.name = t.spb
		WHERE t.receive_type = "TBS Internal" AND t.spb IS NOT NULL AND t.docstatus = 1 {};
	""".format(conditions), filters, as_dict=True)

	return _group_rows(sql)

def get_data_all(filters):
	"""Ambil SEMUA baris tanpa filter t.receive_type sama sekali.
	Grouping ditentukan dari supplier (kalau ada) atau divisi (kalau ada),
	masing-masing diberi prefix supaya jelas asal groupingnya."""
	conditions = get_condition(filters, "All")
	sql = frappe.db.sql("""
		SELECT
			t.name,
			CASE
				WHEN t.supplier IS NOT NULL THEN CONCAT('Supplier: ', s.supplier_name)
				WHEN t.spb IS NOT NULL THEN CONCAT('Divisi: ', spb.divisi)
				ELSE 'Lainnya'
			END as grouping,
			t.no_polisi,
			t.ticket_number as no_referensi,
			t.name as no_slip_timbang,
			t.driver_name as nama_supir,
			t.posting_date as tgl_masuk,
			t.weight_in_time as jam_masuk,
			t.posting_date as tgl_keluar,
			t.weight_out_time as jam_keluar,
			FORMAT(t.bruto, 2) as bruto,
			FORMAT(t.tara, 2) as tara,
			FORMAT(t.netto, 2) as netto_1,
			FORMAT(t.netto * (t.potongan_sortasi / 100), 2) as sort,
			FORMAT(t.netto_2, 2) as netto_2,
			FORMAT(t.isi_komidel, 2) as bjr,
			FORMAT(0, 2) as grader
		FROM `tabTimbangan` t
		LEFT JOIN `tabSupplier` s ON s.name = t.supplier
		LEFT JOIN `tabSurat Pengantar Buah` as spb ON spb.name = t.spb
		WHERE t.receive_type IN ('TBS Internal', 'TBS Eksternal')
			AND (t.supplier IS NOT NULL OR t.spb IS NOT NULL)
			AND t.docstatus = 1 {};
	""".format(conditions), filters, as_dict=True)

	return _group_rows(sql)

def _group_rows(sql_result):
	"""Helper untuk mengelompokkan hasil query berdasarkan kolom 'grouping',
	supaya tidak duplikasi logic defaultdict di 3 fungsi."""
	grouped = defaultdict(list)

	for row in sql_result:
		grouped[row["grouping"]].append(row)

	return [
		{
			"grouping": key,
			"data": value
		}
		for key, value in grouped.items()
	]